rust/gateway/src/main.rs

```diff
@@ -98,39 +100,44 @@ async fn get_firezone_id(env_id: Option<String>) -> Result<String> {
 }
 
 async fn run(login: LoginUrl, private_key: StaticSecret) -> Result<Infallible> {
-    let mut tunnel = GatewayTunnel::new(private_key, Sockets::new(), CallbackHandler)?;
-
-    let (portal, init) = phoenix_channel::init::<_, InitGateway, _, _>(
+    let tunnel = GatewayTunnel::new(private_key, Sockets::new(), CallbackHandler)?;
+    let portal = PhoenixChannel::connect(
         Secret::new(login),
         get_user_agent(None, env!("CARGO_PKG_VERSION")),
         PHOENIX_TOPIC,
         (),
         ExponentialBackoffBuilder::default()
             .with_max_elapsed_time(None)
             .build(),
-    )
-    .await??;
+    );
 
-    tunnel
-        .set_interface(&init.interface)
-        .context("Failed to set interface")?;
-    let mut tun_device = connlib_shared::tun_device_manager::TunDeviceManager::new()?;
-    tun_device
-        .set_ips(init.interface.ipv4, init.interface.ipv6)
-        .await?;
+    let (sender, mut receiver) = mpsc::channel::<Interface>(10);
+
+    let mut tun_device = TunDeviceManager::new()?;
     tun_device
         .set_routes(
             vec![Cidrv4::from(PEERS_IPV4.parse::<Ipv4Network>().unwrap())],
             vec![Cidrv6::from(PEERS_IPV6.parse::<Ipv6Network>().unwrap())],
         )
         .await?;
-    tunnel.update_relays(HashSet::default(), init.relays);
 
-    let mut eventloop = Eventloop::new(tunnel, portal);
+    let update_device_task = async move {
+        while let Some(next_interface) = receiver.next().await {
+            if let Err(e) = tun_device
+                .set_ips(next_interface.ipv4, next_interface.ipv6)
+                .await
```

Comment:
`TunDeviceManager` needs an async context, hence this is running in a separate task.